cloudsync/smartsync.py

```diff
@@ -0,0 +1,186 @@
+                yield ent
+
+    @property
+    def _changeset(self):
+        folders = [ent for ent in self._changeset_storage if ent[REMOTE].otype == DIRECTORY or 
+                   ent[REMOTE].changed > ent[REMOTE]._last_gotten]
+        retval = self.requestset.intersection(self._changeset_storage).union(folders).difference(self.excludeset)
+        return retval
+
+    # def change(self, age):
+    #     super().change(age)
+    #     requested_changes = self.requestset.intersection(self._changeset):
+    #     if requested_changes:
+
+    def unconditionally_get_latest(self, ent, i):
+        # TODO: get enhanced data (mtime, size, etc)
+        retval = super().unconditionally_get_latest(ent, i)
+        return retval
+
+
+class SmartCloudSync(CloudSync):
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 roots: Optional[Tuple[str, str]] = None,
+                 storage: Optional['Storage'] = None,
+                 sleep: Optional[Tuple[float, float]] = None,
+                 root_oids: Optional[Tuple[str, str]] = None
+                 ):
+        super().__init__(providers=providers, roots=roots, storage=storage,
+                         sleep=sleep, root_oids=root_oids, smgr_class=SmartSyncManager, state_class=SmartSyncState)
+
+    def smart_unsync_ent(self, ent: SyncEntry):
+        # remove ent from requestset
+        # delete file, if it exists at path
+        self.state.smart_unsync_ent(ent)
+        if ent[LOCAL].sync_path:
+            ent_info = self.providers[LOCAL].info_path(ent[LOCAL].path)
+            if ent_info:
+                self.providers[LOCAL].delete(ent_info.oid)
+
+    def smart_unsync_path(self, path, side):
+        # TODO: should this take an oid? we may not have a local oid, so it might have to be a remote oid always...
+        path = self._make_path_remote(path, side)
+        ents = self.state.lookup_path(REMOTE, path)
+        if not ents:
+            raise FileNotFoundError(path)
+        for ent in ents:
+            self.smart_unsync_ent(ent)
+
+    def smart_sync_ent(self, ent):
+        # if the local file is missing,
+        #   clear the local side of the ent and
+        #   mark the remote side changed and clear the sync_path/sync_hash so that it will look new and sync down
+        #   this should override the local deletion event from before
+        # TODO: is there a race condition based on the local deletion event coming later? This might wrongly delete!
+        if not self.providers[LOCAL].exists_path(ent[LOCAL].path):
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+            self.state.update_entry(ent, REMOTE, None, changed=True)
+        self.state.smart_sync_ent(ent)
+        for parent_conflict in self.smgr.get_parent_conflicts(ent, REMOTE):  # ALWAYS remote
+            self.smgr.sync_one_entry(parent_conflict)
+        self.smgr.sync_one_entry(ent)
```

Comment:
should probably return if something got done

Reply:
done